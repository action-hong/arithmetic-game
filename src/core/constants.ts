export const TRIES_LIMIT = 6
export const START_DATE = new Date(2022, 3, 7)
export const ANSWER_LENGTH = 8
export const MARKUP = '1234567890+-×÷='.split('')
const SPECIAL_MARKUP = [
  ['*', '×'],
  ['/', '÷'],
]
export const KEY_MARKUP = [
  ...MARKUP.filter(m => m !== '×' && m !== '÷'),
  ...SPECIAL_MARKUP,
]

export const CUSTOM_TRIES_LIMIT_MIN = 3
export const CUSTOM_TRIES_LIMIT_MAX = 10
