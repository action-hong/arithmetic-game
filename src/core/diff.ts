type IndexResultType = 'correct' | 'wrong position' | 'wrong'

interface IndexResult {
  pos: number
  char: string
  type: IndexResultType
}

export function diff(answer: string, input: string): Array<IndexResult> {
  if (answer.length !== input.length) {
    throw new Error('Answer and input must be the same length')
  }
  const result: Array<IndexResult> = []
  const answerChars = answer.split('').map(item => ({
    item,
    selected: false,
  }))
  for (let i = 0; i < answer.length; i++) {
    const char = answer[i]
    const inputChar = input[i]
    if (char === inputChar) {
      result.push({
        pos: i,
        char: inputChar,
        type: 'correct',
      })
      answerChars[i].selected = true
    }
    else if (
      answerChars.findIndex(item => item.item === inputChar && !item.selected) !== -1
    ) {
      result.push({
        pos: i,
        char: inputChar,
        type: 'wrong position',
      })
    }
    else {
      result.push({
        pos: i,
        char: inputChar,
        type: 'wrong',
      })
    }
  }
  return result
}

export function getTipFromResult(result: Array<IndexResult>) {
  return result.reduce((acc, item) => {
    if (!acc.has(item.char)) {
      acc.set(item.char, item.type)
    }
    else {
      const currentType = acc.get(item.char)
      if (currentType !== 'correct' && item.type !== 'wrong') {
        acc.set(item.char, item.type)
      }
    }
    return acc
  }, new Map<string, IndexResultType>())
}
