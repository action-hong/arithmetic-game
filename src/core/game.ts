import type { Options } from 'canvas-confetti'
import confetti from 'canvas-confetti'

export interface TriesMeta {
  answer?: string
  start?: number
  end?: number
  duration?: number
  tries?: string[]
  // 数字位置必须正确10+20=30, 不能是20+10=30
  strict?: boolean
  // 当前的输入, 还没有提交
  currentTry?: string[]
  currentTryIndex?: number
  hint?: boolean
  hintLevel?: number
  passed?: boolean
  failed?: boolean
}

const count = 200
const defaults: Options = {
  origin: { y: 0.7 },
}

function fire(particleRatio: number, opts: Options) {
  confetti(Object.assign({}, defaults, opts, {
    particleCount: Math.floor(count * particleRatio),
  }))
}

export function win() {
  fire(0.25, {
    spread: 26,
    startVelocity: 55,
  })
  fire(0.2, {
    spread: 60,
  })
  fire(0.35, {
    spread: 100,
    decay: 0.91,
    scalar: 0.8,
  })
  fire(0.1, {
    spread: 120,
    startVelocity: 25,
    decay: 0.92,
    scalar: 1.2,
  })
  fire(0.1, {
    spread: 120,
    startVelocity: 45,
  })
}
