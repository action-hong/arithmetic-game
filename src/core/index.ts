export * from './diff'
export * from './game'
