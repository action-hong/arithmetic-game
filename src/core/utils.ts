/**
* Checks whether a given date is in daylight saving time(夏令时).
* @param date the date object to be checked.
* @returns true if the date is in daylight saving time, false if it's not.
*/
export function isDstObserved(date: Date) {
  const jan = new Date(date.getFullYear(), 0, 1)
  const jul = new Date(date.getFullYear(), 6, 1)
  const standardTimezoneOffset = Math.max(jan.getTimezoneOffset(), jul.getTimezoneOffset())
  return date.getTimezoneOffset() < standardTimezoneOffset
}
