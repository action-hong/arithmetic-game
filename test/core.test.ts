import { describe, expect, it } from 'vitest'
import { diff, getTipFromResult } from '~/core/diff'

describe('core', () => {
  it('diff', () => {
    expect(diff(
      '1+1=2',
      '2-1=1',
    )).toMatchSnapshot()

    const result = diff(
      '1+1=234',
      '1+1=147',
    )
    expect(result).toMatchSnapshot()

    expect(getTipFromResult(result)).toMatchSnapshot()
  })
})
