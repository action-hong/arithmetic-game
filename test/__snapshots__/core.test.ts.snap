// Vitest Snapshot v1

exports[`core > diff 1`] = `
[
  {
    "char": "2",
    "pos": 0,
    "type": "wrong position",
  },
  {
    "char": "-",
    "pos": 1,
    "type": "wrong",
  },
  {
    "char": "1",
    "pos": 2,
    "type": "correct",
  },
  {
    "char": "=",
    "pos": 3,
    "type": "correct",
  },
  {
    "char": "1",
    "pos": 4,
    "type": "wrong position",
  },
]
`;

exports[`core > diff 2`] = `
[
  {
    "char": "1",
    "pos": 0,
    "type": "correct",
  },
  {
    "char": "+",
    "pos": 1,
    "type": "correct",
  },
  {
    "char": "1",
    "pos": 2,
    "type": "correct",
  },
  {
    "char": "=",
    "pos": 3,
    "type": "correct",
  },
  {
    "char": "1",
    "pos": 4,
    "type": "wrong",
  },
  {
    "char": "4",
    "pos": 5,
    "type": "wrong position",
  },
  {
    "char": "7",
    "pos": 6,
    "type": "wrong",
  },
]
`;

exports[`core > diff 3`] = `
Map {
  "1" => "correct",
  "+" => "correct",
  "=" => "correct",
  "4" => "wrong position",
  "7" => "wrong",
}
`;
